common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -525,56 +508,50 @@ internal class WebSocketCoreImpl(
                         }
-                }
+                        activeMessages.remove(it.key)
+                    }
             }
         }
         // If we still have flows flowing
-        val hasFlowMessages = synchronized(activeMessages) { activeMessages.any { it.value.eventFlow != null } }
+        val hasFlowMessages = activeMessages.any { it.value.eventFlow != null }
         if (hasFlowMessages && wsScope.isActive) {
             wsScope.launch {
-                // TODO Why a delay ? and why 10s I guess you want to wait for the logic above to finish but 10s is quite arbitrary
-                // You could wait for the job above to finish instead
-                delay(10.seconds)
+                cancelPendingMessagesJob.join()
                 if (connect()) {
                     Timber.d("Resubscribing to active subscriptions...")
-                    // TODO The whole forEach is not synchronized because it sendMessage within but we modify the map while iterating on it
-                    // It never happened because of CRITICAL #2 but it might throw when we are going to remove no?
-                    synchronized(activeMessages) {
-                        activeMessages.filterValues { it.eventFlow != null }.entries
-                    }.forEach { (oldId, oldMessage) ->
-                        // TODO CRITICAL #3 if I'm not wrong we should set back this field to false but I think we should probably create new message instead
-                        // otherwise the two messages stay in activeMessage list (new and old)
-                        // oldMessage.hasContinuationBeenInvoked.set(false)
-
-                        val response = sendMessage(oldMessage)
-                        if (response == null || response.success != true) {
-                            Timber.e("Issue re-registering subscription with ${oldMessage.message}")
-                        } else {
-                            // sendMessage will have created a new item for this subscription
+                    activeMessages.filterValues { it.eventFlow != null }.entries
+                        .forEach { (oldId, oldMessage) ->
                             activeMessages.remove(oldId)
+
+                            val newMessage = WebSocketRequest(
+                                message = oldMessage.message,
+                                eventFlow = oldMessage.eventFlow,
+                                onEvent = oldMessage.onEvent,
+                            )
+
+                            val response = sendMessage(newMessage)
+                            if (response == null || response.success != true) {
+                                Timber.e("Issue re-registering subscription with ${oldMessage.message}")
+                            }
                         }
-                    }
                 } else {
                     // TODO in that case what will happen to the active messages?
```

Comment:
I would like to tackle this in this PR but I would like your opinion on what we should do there @jpelgrom

Reply:
Do you mean retry vs discard? I'm thinking discard, we already did one 'invisible' retry and after that it is a failure for potentially forever so it should be handed back to the part of the app that requested it.

Reply:
If we discard it means that we need to make sure each subscription handle gracefully an exception on the flow. I'm not sure it is the case today but I'll check.

Reply:
Should be tackle in https://github.com/home-assistant/android/issues/5259 I will add it to the TODO in the code.